rustls/src/server/hs.rs

```diff
@@ -72,7 +72,22 @@ pub fn can_resume(
     // establish a new session."
     resumedata.cipher_suite == suite.suite
         && (resumedata.extended_ms == using_ems || (resumedata.extended_ms && !using_ems))
-        && &resumedata.sni == sni
+        && sni_exact_eq(&resumedata.sni, &sni)
+}
+
+/// Exactly compare `DnsName` values
+///
+/// webpki compares `DnsName`s case-insensitively, but for the purpose of comparing SNI values
+/// from resumption to current handshake or HelloRetryRequest to ServerHello, we need an exact
+/// comparison.
+pub(super) fn sni_exact_eq(a: &Option<webpki::DnsName>, b: &Option<webpki::DnsName>) -> bool {
```

Comment:
This function needs to take two `DnsNameRef`s. `DnsName`s are already case-normalized when constructed.